[USMON-1463] Allow CNM & USM to report host/container tags for configurable duration during startup

guyarb opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

Importing the logic of `expected_tags_duration` from the core agent to system-probe (CNM and USM).

Allowing both products to report host/container tags in our payload, to handle race in the backend pipelines.

### Motivation

Today, a if a new host is spawned, there's a race between the pipeline that processes the host and container tags, to the CNM/USM pipeline which tries to pull those tags from the other pipeline.

The issue is "known" in the agent, and handled using a mechaism to report host tags on the relevant metrics for a configurable duration after the startup of the agent.

We import this logic to USM/CNM, and enabling it by default.

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

### Possible Drawbacks / Trade-offs

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

Note - we don't use the core-agent configuration, as it is not enabled by default while the system-probe configuration is.

guyarb committed:
usm,cnm: Report host tags if configured to

guyarb committed:
usm,cnm: Report container tags if available

guyarb committed:
usm,cnm: Add expected_tags_duration configuration

The agent has a configuration called expected_tags_duration, in which we use to report host/container tags
along with our metrics and payloads. We import this behavior to system-probe, without overriding and forcing
the general agent to use a default value

guyarb committed:
usm,cnm: Split taggers

guyarb commented:
Waiting on https://github.com/DataDog/agent-payload/pull/400 to update the go.mods

kkhor-datadog reviewed (commented):
@DataDog/container-experiences shouldn't need to review as our only affected files are in `comp/process/connectionscheck/`, which CNM actually owns the functionality for. But since this is the connections check, Networks should review the changes, not us.

Marking as approved on our end, but please get a review from Networks as well - we can look into updating the CODEOWNERS to reflect the ownership separately.

vickenty commented on pkg/aggregator/host_tag_provider.go:
```diff
@@ -18,23 +18,29 @@ import (
 
-func newHostTagProviderWithClock(clock clock.Clock) *HostTagProvider {
+// NewHostTagProviderWithDuration creates a new HostTagProvider with a specified duration for host tags expiration.
+func NewHostTagProviderWithDuration(duration time.Duration) *HostTagProvider {
```
I think so, there won't be a better time to do the move. Amount of code is small and there aren't much changes to it at the same time.

guyarb replied:
I'll make it in a different PR then

guyarb committed:
usm,cnm: Ensure tag index is always positive

Agent without agent-payload v5.0.162 will provide the default number of the host tags index and local containers tag index
which is 0. However, 0 is a valid tags index, and can cause us to incorrectly add irrelevant tags or just waste CPU and memory
in the backend. Hence, we force the tagger to provide index > 0.

akarpz commented on pkg/config/setup/system_probe.go:
```diff
@@ -312,6 +312,8 @@ func InitSystemProbeConfig(cfg pkgconfigmodel.Config) {
 	cfg.BindEnv(join(netNS, "http_max_request_fragment"))
 	cfg.BindEnv(join(smNS, "http_max_request_fragment"))
 
+	cfg.BindEnvAndSetDefault(join(spNS, "expected_tags_duration"), 10*time.Minute, "DD_SYSTEM_PROBE_EXPECTED_TAGS_DURATION")
```
this seems like a long time? usually the race lasts only a few seconds

akarpz replied:
additionally, what is the the cost of sending the tags from the agent? I assume there must be one if we only do it for a few minutes on startup... is it just customer network from the larger payload size? if the downsides are small, why not just do this forever and remove some backend complexity?

guyarb replied:
> usually the race lasts only a few seconds

That's hard to say, as there's no metric to support that
AFAIK, everytime the core-agent products enable the feature, they do that with 10m or 15m
I'll set it for 5m by default, and in rare cases we can ask the customer to increase it.

> additionally, what is the the cost of sending the tags from the agent? I assume there must be one if we only do it for a few minutes on startup... is it just customer network from the larger payload size? if the downsides are small, why not just do this forever and remove some backend complexity?

The side effects are:
* larger payload of CNM and USM
* CPU consumption by process-agent (fetching the tags)

We should not rely on this as a robust solution and drop backend resolution, as we can spare redudnant calculations for the customer side (The information exists in the backend, it might just take a couple of minutes to fully arrive, parse and index). Also, the customer is able to add user-defined tags from the UI. Those tags won't be available in the agent side.

guyarb committed:
usm,cnm: Change default expected tags duration interval to 5m

guyarb committed:
Merge branch 'main' into arbitman/USMON-1463

guyarb committed:
Merge branch 'main' into arbitman/USMON-1463
